third_party/move/move-compiler/src/typing/expand.rs

```diff
@@ -57,43 +57,98 @@ fn types(context: &mut Context, ss: &mut Vec<Type>) {
+fn unfold_type_or_last_var(subst: &Subst, sp!(loc, t_): Type) -> Result<Type, Spanned<TVar>> {
+    match t_ {
+        Type_::Var(i) => {
+            let last_tvar = forward_tvar(subst, i);
+            match subst.get(last_tvar) {
+                Some(sp!(_, Type_::Var(_))) => panic!("ICE forward_tvar returned a type variable"),
+                Some(inner) => Ok(inner.clone()),
+                None => Err(sp(loc, last_tvar)),
+            }
+        },
+        _ => Ok(sp(loc, t_)),
+    }
+}
+
+// Try to expand the type of ty, warning with msg_uninferred if type cannot be inferred.
+fn type_with_context_msg(context: &mut Context, ty: &mut Type, msg_uninferred: &str) {
     use Type_::*;
     match &mut ty.value {
         Anything | UnresolvedError | Param(_) | Unit => (),
         Ref(_, b) => type_(context, b),
         Var(tvar) => {
             let ty_tvar = sp(ty.loc, Var(*tvar));
-            let replacement = core::unfold_type(&context.subst, ty_tvar);
+            let replacement = unfold_type_or_last_var(&context.subst, ty_tvar);
             let replacement = match replacement {
-                sp!(_, Var(_)) => panic!("ICE unfold_type_base failed to expand"),
-                sp!(loc, Anything) => {
-                    let msg = "Could not infer this type. Try adding an annotation";
+                Err(sp!(loc, last_tvar)) => {
+                    // avoid duplicate error messages
+                    context.subst.insert(last_tvar, sp(ty.loc, Type_::Anything));
                     context
                         .env
-                        .add_diag(diag!(TypeSafety::UninferredType, (ty.loc, msg)));
+                        .add_diag(diag!(TypeSafety::UninferredType, (loc, msg_uninferred)));
                     sp(loc, UnresolvedError)
                 },
-                t => t,
+                Ok(sp!(_, Var(_))) => panic!("ICE unfold_type_base failed to expand"),
+                Ok(t) => t,
             };
             *ty = replacement;
             type_(context, ty);
         },
         Apply(Some(_), sp!(_, TypeName_::Builtin(_)), tys) => types(context, tys),
         Apply(Some(_), _, _) => panic!("ICE expanding pre expanded type"),
-        Apply(None, _, _) => {
+        Apply(None, ty_name, _) => {
+            let ty_name = ty_name.clone();
             let abilities = core::infer_abilities(context, &context.subst, ty.clone());
             match &mut ty.value {
                 Apply(abilities_opt, _, tys) => {
                     *abilities_opt = Some(abilities);
-                    types(context, tys);
+                    for (i, ty) in tys.iter_mut().enumerate() {
+                        type_struct_ty_param(context, ty, i, &ty_name.value)
+                    }
                 },
                 _ => panic!("ICE impossible. tapply switched to nontapply"),
             }
         },
     }
 }
 
+pub fn type_(context: &mut Context, ty: &mut Type) {
```

Comment:
While you're modifying this, can you add a comment to say what this function does?    Perhaps
```
// Try to expand the type of ty, warning with a default message if type cannot be inferred.
```
This whole file is woefully undocumented.  Thanks for figuring it out for us.

Reply:
done